yew/src/virtual_dom/mod.rs

```diff
@@ -60,70 +60,249 @@ impl fmt::Debug for dyn Listener {
+                        .collect(),
+                );
+                match self {
+                    Self::IndexMap(m) => m,
+                    // SAFETY: unreachable because we set the value to the `IndexMap` variant above.
+                    _ => unsafe { unreachable_unchecked() },
+                }
             }
         }
     }
-}
 
-macro_rules! impl_attrs_from {
-    ($($from:path => $variant:ident)*) => {
-        $(
-            impl From<$from> for Attributes {
-                fn from(v: $from) -> Self {
-                    Self::$variant(v)
+    fn diff_vec<'a>(
+        new: &'a [PositionalAttr],
+        old: &[PositionalAttr],
+    ) -> Vec<Patch<&'static str, &'a str>> {
+        let mut out = Vec::new();
+        let mut new_iter = new.iter();
+        let mut old_iter = old.iter();
+
+        loop {
+            match (new_iter.next(), old_iter.next()) {
+                (
+                    Some(PositionalAttr(key, new_value)),
+                    Some(PositionalAttr(old_key, old_value)),
+                ) if key == old_key => match (new_value, old_value) {
+                    (Some(new), Some(old)) => {
+                        if new != old {
+                            out.push(Patch::Replace(*key, new.as_ref()));
+                        }
+                    }
+                    (Some(value), None) => out.push(Patch::Add(*key, value.as_ref())),
+                    (None, Some(_)) => out.push(Patch::Remove(*key)),
+                    (None, None) => {}
+                },
+                // keys don't match, we can no longer compare linearly from here on out
+                (Some(new_attr), Some(old_attr)) => {
+                    // assume that every attribute is new.
+                    let mut added = iter::once(new_attr)
+                        .chain(new_iter)
+                        .filter_map(PositionalAttr::transposed)
+                        .map(|(key, value)| (key, value.as_ref()))
+                        .collect::<HashMap<_, _>>();
+
+                    // now filter out all the attributes that aren't new
+                    for (key, old_value) in iter::once(old_attr)
+                        .chain(old_iter)
+                        .filter_map(PositionalAttr::transposed)
+                    {
+                        if let Some(new_value) = added.remove(key) {
+                            // attribute still exists but changed value
+                            if new_value != old_value.as_ref() {
+                                out.push(Patch::Replace(key, new_value));
+                            }
+                        } else {
+                            // attribute no longer exists
+                            out.push(Patch::Remove(key));
+                        }
+                    }
+
+                    // finally, we're left with the attributes that are actually new.
+                    out.extend(added.into_iter().map(|(k, v)| Patch::Add(k, v)));
```

Comment:
Does it need to be deterministic? It shouldn't matter in which order the patches are applied.
(If it matters then we can't use this algorithm at all because additions will always be performed after `Replace` and `Remove`)

Reply:
Ah, you're right. Order doesn't matter. I was confused 😅